File: old/src/main/kotlin/UI/MainViewModel.cs
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using AutoBuyer.Logic;

namespace AutoBuyer.UI
{
    public class MainViewModel : IBuyerListener, IPortfolioListener
    {
        private readonly List<IUserRequestListener> _listeners;

        public ObservableCollection<BuyerViewModel> Buyers { get; }
        public Command StartBuyingCommand { get; private set; }

        public String NewItemId { get; set; }
        public int NewItemMaximumPrice { get; set; }
        public int NumberToBuy { get; set; }

        public MainViewModel(BuyerPortfolio portfolio)
        {
            StartBuyingCommand = Command(Join);
            Buyers = ObservableCollection<BuyerViewModel>();
            _listeners = List<IUserRequestListener>();
            portfolio.AddPortfolioListener(this);
        }

        private void Join()
        {
            foreach (IUserRequestListener listener in _listeners)
            {
                listener.StartBuying(NewItemId, NewItemMaximumPrice, NumberToBuy);
            }
        }

        public void BuyerStateChanged(BuyerSnapshot snapshot)
        {
            BuyerViewModel viewModel = Buyers.Single(x => x.ItemId == snapshot.ItemId);
            viewModel.UpdateState(snapshot);
        }

        public void BuyerAdded(Buyer buyer)
        {
            var viewModel = BuyerViewModel(buyer.Snapshot.ItemId, buyer.Snapshot);
            Buyers.Add(viewModel);
            buyer.AddBuyerListener(this);
        }

        public void AddUserRequestListener(IUserRequestListener listenter)
        {
            _listeners.Add(listenter);
        }
    }
}

File: old/src/main/kotlin/UI/MainViewModel.kt
class MainViewModel : IBuyerListener, IPortfolioListener {
    private val listeners: List<IUserRequestListener>

    val buyers: ObservableCollection<BuyerViewModel>
    var startBuyingCommand:Command
        private set

    var newItemId: String
    var newItemMaximumPrice: Int
    var numberToBuy: Int

    constructor(portfolio: BuyerPortfolio) {
        startBuyingCommand = Command(Join);
        buyers = ObservableCollection<BuyerViewModel>();
        listeners = List<IUserRequestListener>();
        portfolio.addPortfolioListener(this);
    }

    private fun join() {
        for (listener in listeners) {
            listener.startBuying(newItemId, newItemMaximumPrice, numberToBuy);
        }
    }

    override fun buyerStateChanged(snapshot: BuyerSnapshot) {
        BuyerViewModel viewModel = Buyers . Single (x => x.ItemId == snapshot.ItemId);
        viewModel.UpdateState(snapshot);
    }

    override fun buyerAdded(buyer: Buyer) {
        var viewModel = BuyerViewModel(buyer.snapshot.itemId, buyer.snapshot);
        buyers.Add(viewModel);
        buyer.addBuyerListener(this);
    }

    fun addUserRequestListener(listenter: IUserRequestListener) {
        listeners.add(listenter);
    }
}
}
